File: lib/mercury/hash_cache.rb
require 'sequel'
require 'digest/sha1'

class Mercury::HashCache
  enable_logger

  def initialize(path)
    @file = "#{path}.db"
    @db = Sequel.connect("jdbc:sqlite:#{@file}")
    if @db.table_exists?(:items)
      logger.debug("Items table exists")
    else
      @db.create_table? :items do
        #minimaxi:polsoft:product:1
        primary_key :id
        String :tenant
        String :source
        String :resource
        String :key
        #md5
        String :data_hash
        String :data_size
        Timestamp :created_at
        Timestamp :updated_at
        Timestamp :last_hit_at
        index [:tenant, :source, :resource, :key], unique: true
      end
    end
    @items = @db[:items]
  end

  # FIXME transaction
  def hit?(tenant:, source:, resource:, key:, data:, &block)
    recs = @items.where(tenant: tenant, source: source, resource: resource, key: key)
    rec = recs.first
    hash = Digest::SHA1.hexdigest(data || "")
    now = Time.now
    # FIXME upsert! https://stackoverflow.com/questions/60287095/correct-usage-of-the-sqlite-on-conflict-clause#60287672
    # http://sequel.jeremyevans.net/rdoc-adapters/classes/Sequel/SQLite/DatasetMethods.html#method-i-insert_conflict
    if rec
      if hash == rec[:data_hash]
        logger.debug("Hit #{rec}")
        recs.update(last_hit_at: now)
        return true
      else
        yield
        recs.update(data_hash: hash, data_size: data.size, updated_at: now)
        return false
      end
    else
      yield
      @items.insert(tenant: tenant, source: source, resource: resource, key: key,
                    created_at: now, updated_at: now, data_hash: hash, data_size: data.size)
      return false
    end
  rescue => e
    logger.error "Failed cache #{e.inspect}"
  end

  def close
    @db.disconnect if @db.valid_connection?
    @db = nil
  end

  def drop!
    close
    logger.warn("Deleting #{@file}")
    File.delete(@file)
  end
end

File: src/main/java/pl/amitec/mercury/persistence/HashCache.java
package pl.amitec.mercury.persistence;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.jdbc.core.JdbcTemplate;
import org.sqlite.SQLiteDataSource;
import pl.amitec.mercury.MercuryException;
import pl.amitec.mercury.util.Utils;

import java.nio.file.Paths;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class HashCache implements Cache {

    private static final Logger logger = LoggerFactory.getLogger(HashCache.class);

    private JdbcTemplate jdbcTemplate;
    private String file;

    public HashCache(String path) {
        this.file = path;
        SQLiteDataSource dataSource = new SQLiteDataSource();
        //ds.setLogWriter();
        dataSource.setUrl("jdbc:sqlite:" + file);
        this.jdbcTemplate = new JdbcTemplate(dataSource);

        if (!tableExists()) {
            createTable();
        }
    }

    @Override
    public <E> boolean hit(String tenant, String source, String resource, String key, E data, Consumer<E> dataConsumer) {
        String strData = data.toString();
        String hash = Utils.sha1HexDigest(strData != null ? strData : "");
        Timestamp now = new Timestamp(System.currentTimeMillis());

        List<Map<String, Object>> result = jdbcTemplate.queryForList(
                "SELECT * FROM items WHERE tenant = ? AND source = ? AND resource = ? AND key = ?",
                tenant, source, resource, key);

        if (result != null && !result.isEmpty()) {
            var rec = result.getFirst();
            if (hash.equals(rec.get("data_hash"))) {
                logger.debug("Hit {}", rec);
                jdbcTemplate.update(
                        "UPDATE items SET last_hit_at = ? WHERE tenant = ? AND source = ? AND resource = ? AND key = ?",
                        now, tenant, source, resource, key);
                return true;
            } else {
                dataConsumer.accept(data);
                jdbcTemplate.update(
                        "UPDATE items SET data_hash = ?, data_size = ?, updated_at = ? WHERE tenant = ? AND source = ? AND resource = ? AND key = ?",
                        hash, strData.length(), now, tenant, source, resource, key);
                return false;
            }
        } else {
            dataConsumer.accept(data);
            jdbcTemplate.update(
                    "INSERT INTO items (tenant, source, resource, key, created_at, updated_at, data_hash, data_size)" +
                            " VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
                    tenant, source, resource, key, now, now, hash, strData.length());
            return false;
        }
    }

    private boolean tableExists() {
        Integer count = jdbcTemplate.queryForObject(
                "SELECT COUNT(*) FROM sqlite_master WHERE type='table' AND name='items'",
                Integer.class);
        return count != null && count > 0;
    }

    private void createTable() {
        jdbcTemplate.execute(
                "CREATE TABLE items (" +
                        "id INTEGER PRIMARY KEY, " +
                        "tenant TEXT, " +
                        "source TEXT, " +
                        "resource TEXT, " +
                        "key TEXT, " +
                        "data_hash TEXT, " +
                        "data_size TEXT, " +
                        "created_at TIMESTAMP, " +
                        "updated_at TIMESTAMP, " +
                        "last_hit_at TIMESTAMP, " +
                        "UNIQUE(tenant, source, resource, key))");
    }

    @Override
    public void close() {
        try {
            jdbcTemplate.getDataSource().getConnection().close();
        } catch (SQLException e) {
            throw new MercuryException("DB closing problem", e);
        }
    }

    @Override
    public void drop() {
        close();
        logger.warn("Deleting {}", file);
        Paths.get(file).toFile().delete();
    }
}